compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -243,6 +243,24 @@ pub(crate) fn turn_into_const_value<'tcx>(
     op_to_const(&ecx, &mplace.into(), /* for diagnostics */ false)
 }
 
+/// Computes the tag (if any) for a given type and variant.
+#[instrument(skip(tcx), level = "debug")]
+pub fn tag_for_variant_provider<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    (ty, variant_index): (Ty<'tcx>, abi::VariantIdx),
+) -> Option<ty::ScalarInt> {
+    assert!(ty.is_enum());
+
+    let ecx = InterpCx::new(
```

Comment:
```suggestion
    // `tag_for_variant` needs to do some machine arithmetic, let's create an interpreter instance for that.
    let ecx = InterpCx::new(
```

Ideally we could do that arithmetic without a full `ecx`. In fact maybe you should do what the MIR opts do and use the `DummyMachine` rather than the `CompileTimeInterpreter`?

Does this need to be a query? If rustc_transmute can depend on rustc_const_eval then this could be just a little helper function defined locally in rustc_transmute.

Reply:
> Does this need to be a query? If rustc_transmute can depend on rustc_const_eval then this could be just a little helper function defined locally in rustc_transmute.

Unfortunately, `rustc_transmute` cannot depend on `rustc_const_eval` without creating a dependency cycle.

Reply:
specifically: `rustc_transmute` -> `rustc_const_eval` -> `rustc_trait_selection` -> `rustc_transmute`

Reply:
> Ideally we could do that arithmetic without a full `ecx`. In fact maybe you should do what the MIR opts do and use the `DummyMachine` rather than the `CompileTimeInterpreter`?

`rustc_mir_transform` (which defines `DummyMachine`) already depends on `rustc_const_eval`, so doing so would create a dependency cycle. I'm not sure how much work it'd be to lift `DummyMachine` into `rustc_const_eval`.

Reply:
Shouldn't be very difficult to pull that up to some module in `rustc_const_eval`, though I'll leave it up to Ralf to decide whether that's something that should be done.

Reply:
It was very easy to pull `DummyMachine` up into `rustc_const_eval`, so I've gone ahead and done it. Happy to revert if there are objections.